Update ic-cdk usage to eliminate deprecation warnings when using ic-cdk 0.18.3 in other projects

552020 opened the pull request:
Hallo! 

As you know we are working on the expansion of the **file_to_eth_demo**. Moving from the `chainkey_testing_canister` to the vetKeys management canister, we upgraded the `ic-cdk`  from `0.17.1` to `0.18.3` to access the new vetKeys APIs. See Changelog --> https://github.com/dfinity/cdk-rs/blob/0.18.3/ic-cdk/CHANGELOG.md

---
**The Problem**

After upgrading to `ic-cdk = "0.18.3"`, we got deprecation warnings from the auto-generated bindings for `ic_siwe_provider`. While these are just warnings and don't break functionality, they create unnecessary noise in build outputs.


The warnings were:

```
warning: use of deprecated type alias `ic_cdk::api::call::CallResult`: Please use `ic_cdk::call::CallResult` instead.
warning: use of deprecated function `ic_cdk::call`: Please use `ic_cdk::call::Call::unbounded_wait()` instead.
```

These warnings are caused by the auto-generated `ic_siwe_provider.rs` file that `ic-cdk-bindgen` creates from the `.did` file.

---
**The Cause**

The `ic_siwe` and `ic_siwe_provider` canisters were using an older version of `ic-cdk` (`0.15.1`) that contained deprecated APIs:

- `ic_cdk::spawn` (deprecated)
- `ic_cdk::call` (deprecated)
- `ic_cdk::api::set_certified_data` (deprecated)
- `ic_cdk::caller` (deprecated)
- `ic_cdk::id` (deprecated)

The modern APIs are:

- `ic_cdk::futures::spawn`
- `ic_cdk::call::Call::unbounded_wait()`
- `ic_cdk::api::certified_data_set`
- `ic_cdk::api::msg_caller`
- `ic_cdk::api::canister_self`

___
**The Solution**

We updated the dependencies in `Cargo.toml` and then vibe solved the warnings one by one:

- 1. Updated Dependencies

```toml
[dependencies]
ic-cdk = "0.18.3"  # Updated from 0.15.1
```

- 2. Fixed Each Warning

**In `ic_siwe/src/init.rs`:**

```rust
// Before
ic_cdk::spawn(async {
    let (seed,): ([u8; 32],) = ic_cdk::call(Principal::management_canister(), "raw_rand", ()).await.unwrap();
});

// After
ic_cdk::futures::spawn(async {
    let response = ic_cdk::call::Call::unbounded_wait(Principal::management_canister(), "raw_rand")
        .with_arg(())
        .await
        .unwrap();
    let (seed,): ([u8; 32],) = candid::decode_one(&response.into_bytes()).unwrap();
});
```

**In `ic_siwe_provider/src/lib.rs`:**

```rust
// Before
use ic_cdk::api::set_certified_data;
set_certified_data(&prefixed_root_hash[..]);

// After
use ic_cdk::api::certified_data_set;
certified_data_set(&prefixed_root_hash[..]);
```

**In `ic_siwe_provider/src/service/get_caller_address.rs`:**

```rust
// Before
let principal = ic_cdk::caller();

// After
let principal = ic_cdk::api::msg_caller();
```

**In `ic_siwe_provider/src/service/init_upgrade.rs` and `siwe_login.rs`:**

```rust
// Before
let canister_id = ic_cdk::id();
&ic_cdk::api::id()

// After
let canister_id = ic_cdk::api::canister_self();
&ic_cdk::api::canister_self()
```
---

**Warning**

It builds but it has not been really tested. So this PR is more an Issue with the proposal of a solution.

552020 committed:
chore: update ic-cdk to 0.18.3 in ic_siwe and ic_siwe_provider

552020 committed:
refactor(ic_siwe): replace deprecated ic_cdk::spawn and ic_cdk::call with modern API

552020 committed:
refactor(ic_siwe_provider): replace deprecated set_certified_data with certified_data_set

552020 committed:
refactor(ic_siwe_provider): replace deprecated ic_cdk::caller with msg_caller

552020 committed:
refactor(ic_siwe_provider): replace deprecated ic_cdk::id with canister_self

kristoferlund commented:
Nice @552020! Yes, some upgrades needs to be done, both for the SIWE provide and library as well as the frontend libraries. I will merge this to a dev branch and include in next release. Thanks!

552020 committed:
fix: update ic-cdk dependencies to resolve version conflicts

- Update ic-cdk-timers from 0.9.0 to 0.12.2 in ic_siwe package
- Update ic-cdk from 0.15.0 to 0.18.3 in test_canister package
- Resolves compatibility issues with ic-cdk 0.18.3

552020 committed:
fix(rand): prevent panic when RNG is not yet initialized

This patch prevents a runtime panic caused by calling `unwrap()` on an uninitialized RNG instance in `generate_nonce()`.

The original implementation assumed that the RNG (thread-local) would always be initialized before `prepare_login()` is invoked. However, on the Internet Computer, RNG initialization is typically triggered asynchronously via timers — meaning it may not be ready when `prepare_login()` is first called.

This change:

- Updates `generate_nonce()` to return `Result<String, String>` instead of panicking.
- Returns an error when RNG is not yet initialized, allowing calling functions to fail gracefully.
- Modifies `prepare_login()` to propagate the error by mapping it to `EthError::Other`.

This approach avoids IC0503 canister trap errors and allows frontend clients to retry login preparation gracefully after a short delay, improving developer experience and system stability.

Note: This does **not** change the test behavior. In `#[cfg(test)]`, `generate_nonce()` still uses `thread_rng()` and always succeeds.

Follow-up improvements may include:
- Forcing RNG initialization in `init()` using a dummy `raw_rand` request.
- Exposing RNG readiness as part of the library state.

Fixes: `Panicked at 'called Option::unwrap() on a None value', rand.rs:7`

552020 committed:
feat(eth): add Other variant to EthError for RNG error handling

552020 committed:
refactor: replace thread_local with OnceCell and RwLock for IC compatibility

- Replace thread_local! with OnceCell<RwLock<...>> for global state
- Add ensure_globals_initialized() function for safe initialization
- Update all modules to use RwLock-based access patterns
- Fix temporary value borrowing issue in macros.rs
- Add once_cell dependency to Cargo.toml

This addresses the root cause of RNG initialization failures on the Internet Computer.

552020 committed:
fix: resolve init mode error and remove unused imports

- Remove immediate ic_cdk::futures::spawn() call from init_rng() to avoid 'ic0_call_new cannot be executed in init mode' error
- Keep only timer-based RNG initialization to comply with IC init mode rules
- Remove unused SETTINGS import from init.rs
- Remove unused SIWE_MESSAGES import from login.rs
- Library now compiles without warnings and can be deployed successfully

552020 committed:
fix: correct Candid decoding for raw_rand response

- Change from [u8; 32] to Vec<u8> for raw_rand response decoding
- Convert Vec<u8> to [u8; 32] array for ChaCha20Rng seed
- This fixes the 'Fail to decode argument 0' error during RNG initialization

552020 committed:
feat: add once_cell dependency for global state management

- Add once_cell = 1.19.0 to Cargo.toml
- Update Cargo.lock with new dependency
- This enables safe global state initialization for IC compatibility

552020 committed:
fix: correct Candid decoding for raw_rand response and ignore IC-specific doctest

- Change from [u8; 32] to Vec<u8> for raw_rand response decoding
- Convert Vec<u8> to [u8; 32] array for ChaCha20Rng seed
- Change doctest from ignore to prevent IC-specific test failures
- This fixes the 'Fail to decode argument 0' error during RNG initialization

552020 committed:
fix: update delegation tests to use new global state pattern

- Update test initialization to use ensure_globals_initialized() and RwLock pattern
- Remove unused SETTINGS import to eliminate compiler warnings
- Fix test compatibility with OnceCell-based global state management

552020 committed:
feat: add test module to lib.rs

- Add #[cfg(test)] mod tests; to enable the new test module
- This allows the comprehensive test suite to be included in the library

552020 committed:
test: add comprehensive test suite

- Add tests.rs with comprehensive test coverage
- Test global state initialization
- Test nonce generation in test mode
- Test settings builder creation
- Test Ethereum address creation and validation
- All tests pass without warnings
- Provides better coverage for the refactored library

552020 committed:
fix: correct raw_rand response decoding format

- Change from (Vec<u8>,) tuple to Vec<u8> direct decoding
- raw_rand returns Vec<u8> directly, not wrapped in a tuple
- This fixes the 'Fail to decode argument 0' error during RNG initialization

552020 committed:
feat: add enhanced debug logging and RNG readiness check for IC compatibility

kristoferlund commented:
Hi @552020! Thanks for the ambitious work you are doing. It seems though that this PR now is covering more than its initial scope? For me to understand the reasons for all the changes, please keep PRs small and well described.

552020 commented:
Oh sorry! All this 'work' (mess) was not meant to be pushed into this PR. I was trying to integrate the library into my project and was making some tweaks for personal use trying to solve some problems. I was not aware that the changes would have been pushed into this PR. I'll close this PR and open another one just with the first commit. :-)

kristoferlund commented:
@552020, I have time this week to upgrade ic-siwe. If you will, please push those initial commits again as a separate PR like you mentioned.
